pkg/controller/master/setting/containerd_registry.go

```diff
@@ -0,0 +1,154 @@
+				Name:      util.ContainerdRegistrySecretName,
+				Namespace: util.CattleSystemNamespaceName,
+			},
+			Data: map[string][]byte{
+				util.ContainerdRegistryFileName: registryFromSettingYaml,
+			},
+		}
+
+		if _, err := h.secrets.Create(newSecret); err != nil {
+			return err
+		}
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	isStale, err := h.isStaleSecret(secret, registryFromSetting)
+	if err != nil {
+		return err
+	}
+
+	if isStale {
+		toUpdate := secret.DeepCopy()
+		toUpdate.Data = map[string][]byte{
+			util.ContainerdRegistryFileName: registryFromSettingYaml,
+		}
+		if _, err := h.secrets.Update(toUpdate); err != nil {
+			return err
+		}
+	}
+
+	if err = h.removeCredentialInSetting(setting, registryFromSetting); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (h *Handler) clearContainerdRegistrySecret() error {
+	secret, err := h.secretCache.Get(util.CattleSystemNamespaceName, util.ContainerdRegistrySecretName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	toUpdate := secret.DeepCopy()
+	toUpdate.Data = map[string][]byte{
+		util.ContainerdRegistryFileName: []byte(""),
+	}
+	if !reflect.DeepEqual(secret, toUpdate) {
+		if _, err := h.secrets.Update(toUpdate); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (h *Handler) isStaleSecret(secret *corev1.Secret, registryFromSetting *registries.Registry) (bool, error) {
+	isStale := secret.Data == nil || string(secret.Data[util.ContainerdRegistryFileName]) == ""
+	if !isStale {
+		registryFromSecret := &registries.Registry{}
+		if err := yaml.Unmarshal(secret.Data[util.ContainerdRegistryFileName], registryFromSecret); err != nil {
+			return true, err
+		}
+		// We only compare mirros, because we will remove configs.
+		isStale = isStale || !isSameMirrors(registryFromSetting.Mirrors, registryFromSecret.Mirrors)
+	}
+
+	return isStale, nil
+}
+
+func (h *Handler) removeCredentialInSetting(setting *harvesterv1.Setting, registry *registries.Registry) error {
+	// Remove credential in configs.
+	for configName, config := range registry.Configs {
+		config.Auth = nil
+		registry.Configs[configName] = config
+	}
+
+	registryJson, err := json.Marshal(registry)
+	if err != nil {
+		return err
+	}
+
+	toUpdate := setting.DeepCopy()
+	toUpdate.Value = string(registryJson)
+	if !reflect.DeepEqual(setting, toUpdate) {
+		if _, err = h.settings.Update(toUpdate); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func isSameMirrors(fromSetting, fromSecret map[string]registries.Mirror) bool {
+	if len(fromSetting) == 0 && len(fromSecret) == 0 {
+		return true
+	}
+
+	if len(fromSetting) == 0 || len(fromSecret) == 0 {
+		return false
+	}
+
+	// For map comparison in reflect.DeepEqual, it only same if both are nil or both content are same.
+	for mirrorName, mirror := range fromSetting {
+		if len(mirror.Rewrites) == 0 {
+			mirror.Rewrites = nil
+		}
+		fromSetting[mirrorName] = mirror
```

Comment:
before:
```
		if len(mirror.Rewrites) == 0 {
			mirror.Rewrites = nil
		}
		fromSetting[mirrorName] = mirror
```

after
```
		if len(mirror.Rewrites) == 0 && mirror.Rewrites != nil {
			mirror.Rewrites = nil
			fromSetting[mirrorName] = mirror
		}
```
should it be such ?

Reply:
Fix in https://github.com/harvester/harvester/pull/2830. Thanks.